### PR TITLE
for #11830 added new metric for collecting startup method from all startup phases

### PR DESCRIPTION
**Pull request by sraturi**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by sraturi** [comment]
### Test Flows
 

- Start app from app icon (cold startup) -> press home button ->press the app icon again(warm startup)

- Start app from app icon (cold startup) -> press home button ->open Gmail link [open to browser option](warm startup)
    - Recording telemetry inside onStart was using the HomeActivty's existing intent to get the startup information. For example, on cold startup, it would report APP_ICON and than on warm startup(Gmail link) it would again report APP_ICON. Therefore, had to record the telemetry inside the `onNewIntent()` to report the correct startup method, LINK in this case.

- open Gmail link[in app browser](cold startup)-> press home button -> open twitter link [in app browser] (cold startup)
     - This always begins with a cold startup... I'm assuming because Each time, we create a new `ExternalAppBrowserActivity` instance in the respective in-app processes?

- open Gmail link [open to browser option](cold startup) -> press home button ->open twitter link [open to browser option](warm startup)

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 1]
```diff
@@ -308,6 +323,15 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         }
     }
 
+    @VisibleForTesting(otherwise = PROTECTED)
+    internal open fun getIntentStartSource(intent: SafeIntent): Event.OpenedAppAllStart.Source? {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:327]: nit: `internal` has no effect on visibility if you're not being consumed by another build module so we can remove this

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 2]
```diff
@@ -308,6 +323,15 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         }
     }
 
+    @VisibleForTesting(otherwise = PROTECTED)
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:326]: nit: I don't see this being used for testing: shall we make this method `protected` and skip the annotation?

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 1 above)
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:327]: nit: `AllStartSource`

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 2 above)
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:326]: Ah yeah, most of the functions was copied from the existing code for "app_opened" metric. I was wondering if unit testing needs to be part of this PR?

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 3]
```diff
@@ -229,6 +234,16 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         val intentHandled = intentProcessors.any { it.process(intent, navHost.navController, this.intent) }
         browsingModeManager.mode = getModeFromIntentOrLastKnown(intent)
 
+        if (settings().isTelemetryEnabled) {
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:237]: Gotcha!

**Comment by sraturi** [comment]
# Request for data collection review form

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

**1.  What questions will you answer with this data?**
		How do users open the app (custom tab, app_icon, link) most frequently on all startups (cold, warm, hot)

**2.  Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:**
	Creating accurate understanding of real world usage to prioritize performance improvements  
  

**3.  What alternative methods did you consider to answer these questions? Why were they not sufficient?**
    There is an existing app_opened metrics, However it is only used for cold startup.  
  

**4.  Can current instrumentation answer these questions?**
    no  
      
    

5.  **List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.**
   data is Interaction data.How the user opens app , options -> custom tab, app_icon, link  
  

**6.  Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.**
   This is documented in the metrics.md file

**7.  How long will this data be collected?**
    Until 2020-12-01  
  

**8.  What populations will you measure?**
All release, beta, and nightly users with telemetry enabled.  
  

**9.  If this data collection is default on, what is the opt-out mechanism for users?**
default on, opt out by Settings ->Data collection

**10.  Please provide a general description of how you will analyze this data.**
analyze so we can set up better tests and prioritize analyses to improve performance in the most impactful areas.

**11.  Where do you intend to share the results of your analysis?**
Glean, Amplitude and with mobile teams.  
  

**12.  Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:**
N/A

@liuche

File: app/metrics.yaml [hunk 4]
```diff
@@ -8,6 +8,23 @@ no_lint:
   - CATEGORY_GENERIC
 
 events:
+  app_opened_all_startup:
+    type: event
+    description: |
+      A user opened the app from any one of the following flow
+      hot, warm or cold start to the homescreen.
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done.

File: app/metrics.yaml [hunk 5]
```diff
@@ -8,6 +8,23 @@ no_lint:
...
+      source:
+        description: |
+          The method used to open Fenix. Possible values are `app_icon`,
+          `custom_tab` or `link`
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/11830
+    data_reviews:
+      - NA
+    notification_emails:
+      - fenix-core@mozilla.com
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 6]
```diff
@@ -229,6 +234,16 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
...
+            /* If there is a warm or hot startup, onNewIntent method is always called first.
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: done

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt [hunk 7]
```diff
@@ -263,6 +263,12 @@ sealed class Event {
         override val extras: Map<Events.appOpenedKeys, String>?
             get() = hashMapOf(Events.appOpenedKeys.source to source.name)
     }
+    data class OpenedAppAllStart(val source: Source) : Event() {
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt:267]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 1 above)
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: done

Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 8]
```diff
@@ -308,6 +323,15 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         }
     }
 
+    @VisibleForTesting(otherwise = PROTECTED)
+    internal open fun getIntentStartSource(intent: SafeIntent): Event.OpenedAppAllStart.Source? {
+        return when {
+            intent.isLauncherIntent -> Event.OpenedAppAllStart.Source.APP_ICON
+            intent.action == Intent.ACTION_VIEW -> Event.OpenedAppAllStart.Source.LINK
+            else -> null
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:328]: Will update it!

[the remark this answers is not in the document]
Comment (MarcLeclair) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:328]: @mcomella  if we record unknown, should we at least store the value of the "UNKNOWN" so that we have some insight on what those different starts were in order to adapt the code afterwards?

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: I added a new class `AllSourceStartupTelemetry` to handle the logic of triggering recording our startup telemetry.

File: app/metrics.yaml [hunk 9]
```diff
@@ -8,6 +8,23 @@ no_lint:
...
+      hot, warm or cold start to the HomeActivity.
+    extra_keys:
+      source:
+        description: |
+          The method used to open Fenix. Possible values are `app_icon`,
+          `custom_tab` or `link`
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/11830
+    data_reviews:
+      - NA
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done

File: app/metrics.yaml [hunk 10]
```diff
@@ -8,6 +8,23 @@ no_lint:
   - CATEGORY_GENERIC
 
 events:
+  app_opened_all_startup:
+    type: event
+    description: |
+      A user opened the app from any one of the following flows:
+      hot, warm or cold start to the HomeActivity.
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done

File: app/metrics.yaml [hunk 11]
```diff
@@ -8,6 +8,23 @@ no_lint:
...
+      source:
+        description: |
+          The method used to open Fenix. Possible values are `app_icon`,
+          `custom_tab` or `link` or `unknown`
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/11830
+    data_reviews:
+      - https://github.com/mozilla-mobile/fenix/pull/11940/
+    notification_emails:
+      - esmyth@mozilla.com
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml:28]: done

File: app/metrics.yaml [hunk 12]
```diff
@@ -8,6 +8,23 @@ no_lint:
...
+  app_opened_all_startup:
+    type: event
+    description: |
+      A user opened the app from any one of the following flows:
+      hot, warm or cold start to the HomeActivity.
+    extra_keys:
+      source:
+        description: |
+          The method used to open Fenix. Possible values are `app_icon`,
+          `custom_tab` or `link` or `unknown`
```
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done

File: app/metrics.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (sraturi) [reply app/metrics.yaml]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 13]
```diff
@@ -133,7 +133,6 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
     }
 
     private lateinit var navigationToolbar: Toolbar
-
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:136]: nit: try not to remove unrelated lines :)

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 14]
```diff
@@ -250,6 +253,14 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
                 ?.let { it as? TabTrayDialogFragment }
                 ?.also { it.dismissAllowingStateLoss() }
         }
+        // If there is a warm or hot startup, onNewIntent method is always called first.
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:256]: nit: newline above. You should space related code in blocks detached with newlines from other blocks

File: app/metrics.yaml [hunk 15]
```diff
@@ -8,6 +8,26 @@ no_lint:
   - CATEGORY_GENERIC
 
 events:
+  app_opened_all_startup:
+    type: event
+    description: |
+      A user opened the app to the HomeActivity. The HomeActivity
```
Comment (mcomella) [inline app/metrics.yaml:14]: "The system received an Intent for the HomeActivity. An intent is received an external entity wants to the app to display content. Intents can be received when the app is closed – at which point the app will be opened – or when the app is already opened – at which point the already open app will make changes such as loading a url. This can be used loosely as a heuristic for when the user requested to open the app. The HomeActivity encompasses..."

Or something like that. We need to make it clear this is not "user opened the app" anymore.

File: app/metrics.yaml [hunk 16]
```diff
@@ -8,6 +8,26 @@ no_lint:
   - CATEGORY_GENERIC
 
 events:
+  app_opened_all_startup:
```
Comment (mcomella) [inline app/metrics.yaml:11]: We have to change the name to reflect that this is opening intents: `app_received_intent`?

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 17]
```diff
@@ -320,6 +331,14 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         }
     }
 
+    open fun getIntentAllSource(intent: SafeIntent): Event.OpenedAppAllStart.Source? {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:334]: nit: protected

**Comment by mcomella** [comment]
We should get another data review but let's not block the PR on that.

File: app/metrics.yaml (hunk 15 above)
Comment (liuche) [reply app/metrics.yaml:14]: Honestly, I would put all those details below what was initially written. Let's get the simple version first so it's easy for someone to understand the high-level, and then move all the details afterwards.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 17 above)
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 14 above)
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:258]: done

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 13 above)
Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt]: my bad!

File: app/metrics.yaml (hunk 16 above)
Comment (sraturi) [reply app/metrics.yaml]: done!

File: app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt [hunk 18]
```diff
@@ -296,6 +296,13 @@ sealed class Event {
             get() = hashMapOf(Events.appOpenedKeys.source to source.name)
     }
 
+    data class AppRecievedIntent(val source: Source) : Event() {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt:299]: nit: can we correct the spelling, in follow-up to avoid re-running CI? Should be `Received`. Here and in several other places

Comment (sraturi) [reply app/src/main/java/org/mozilla/fenix/components/metrics/Metrics.kt:299]: sorry I tried hard to teach myself its "e" before "i"... but starting to think there is a bigger issue here 😅

File: app/metrics.yaml (hunk 15 above)
Comment (mcomella) [reply app/metrics.yaml]: @sraturi Can you address this in the follow-up? It should be okay to change the description

**Comment by sraturi** [comment]
# Request for data collection review form  

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

**1.  What questions will you answer with this data?**
		What type of intent does the HomeActivity receive most frequently?  for example, an intent with ACTION_VIEW, ACTION_MAIN, launcher intent, etc. can also be used loosely as a heuristic for when the user requested to open the app.

**2.  Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:**
	Creating an accurate understanding of real-world usage to prioritize performance improvements  
  

**3.  What alternative methods did you consider to answer these questions? Why were they not sufficient?**
    There is an existing app_opened metrics, However it is only used for cold startup.  
  

**4.  Can current instrumentation answer these questions?**
    no  
      
    

**5.  List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki.**
data is Interaction data. How the user opens app , options -> custom tab, app_icon, link. Determined based on the intent received by the home activity
  

**6.  Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.**
   This is documented in the metrics.md file

**7.  How long will this data be collected?**
    Until 2020-12-01  
  

**8.  What populations will you measure?**
All release, beta, and nightly users with telemetry enabled.  
  

**9.  If this data collection is default on, what is the opt-out mechanism for users?**
default on, opt out by Settings ->Data collection

**10.  Please provide a general description of how you will analyze this data.**
analyze so we can set up better tests and prioritize analyses to improve performance in the most impactful areas.

**11.  Where do you intend to share the results of your analysis?**
Glean, Amplitude and with mobile teams.  
  

**12.  Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:**
N/A


@liuche another data collection review form to describe the data we are collecting more accurately
